ruby/phase-1.md

```diff
@@ -0,0 +1,575 @@
+
+#### Fibers
+
+Ruby also has Fibers that allow light weight cooperative concurrency. These are code blocks that can
+be paused (from inside) and resumed (from outside). These are really useful for controlling the
+execution of workflow code. However unlike threads Fibers are never preempted and their scheduling
+has to be done manually by the code:
+
+```ruby
+fiber = Fiber.new do
+  i = 0
+  loop do
+    i += 1
+    puts i
+    Fiber.yield
+  end
+end
+
+fiber.resume
+# => 1
+fiber.resume
+# => 2
+fiber.resume
+# => 3
+```
+
+*NOTE: Each `fiber.resume` call blocks the main thread until a Fiber `yield`s back control.*
+
+#### Reactors
+
+Somewhat less popular, but still very viable are libraries that implement a reactor pattern for
+building concurrent applications (e.g. Async or EventMachine). These libraries allow users to
+concurrently execute IO operations.
+
+There are however quite a few big limitations to this approach:
+
+- You need to use very specific libraries to perform IO calls this way (there isn't one for gRPC)
+- Any CPU-bound instructions block the reactor and therefore need to be executed in small bursts per
+  reactor tick
+- Any user-defined code can block the whole thing and has to be wrapped in threads anyways
+
+
+## A basic end-to-end example
+
+Perhaps a bit unconventionally (as far as the proposal review process goes), but we'd like to start
+with an end-to-end example of a simple and fully functional "Hello World" workflow. The intention of
+this example is to demonstrate how everything ties together before diving into each separate moving
+part and exploring it in depth.
+
+```ruby
+require 'temporal'
+require 'temporal/worker'
+
+class MyActivity < Temporal::Activity
+  def execute(name)
+    puts "Hello World, #{name}!"
+  end
+end
+
+class MyWorkflow < Temporal::Workflow
+  def execute(name)
+    # This part is outside of the scope of the 1st phase of the proposal
+    MyActivity.execute!(name, options: { task_queue: 'my-queue' })
+  end
+end
+
+config = Temporal::Configuration.new(
+  url: 'localhost:7233'
+)
+
+client = Temporal::Client.new(config)
+client.start_workflow(MyWorkflow, 'Alice', options: {
+  namespace: 'my-namespace',
+  task_queue: 'my-queue'
+})
+
+worker = Temporal::Worker.new(config, 'my-namespace', 'my-queue')
+worker.register_workflow(MyWorkflow)
+worker.register_activity(MyActivity)
+worker.run
+```
+
+
+## Client
+
+The client is the primary point for interacting with workflows and async activities as well as
+managing namespaces. A client is created with a configuration object that specifies options shared
+between a worker and a client.
+
+```ruby
+config = Temporal::Configuration.new(url: 'localhost:7233')
+client = Temporal::Client.new(config)
+client.start_workflow(MyWorkflow, 'arg_1', kwarg_2: 'foo', options: {
+  namespace: 'my-namespace',
+  task_queue: 'my-queue'
+})
+```
+
+Notes:
+
+- Client is intended to be multi-tenant capable (no limit on a single namespace)
```

Comment:
It's a valid point. The main reason for not enforcing it here is because I have a plan for phase 2 that abstracts namespace and task queue plus other options and therefore makes everything connect seamlessly. But for that I need a more explicit interface underneath. Let's discuss this further, I think everyone has pointed it out and I'd like to get my point across